rules/aip0136/response_message_name.go

```diff
@@ -0,0 +1,64 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aip0136
+
+import (
+	"fmt"
+
+	"bitbucket.org/creachadair/stringset"
+	"github.com/googleapis/api-linter/lint"
+	"github.com/googleapis/api-linter/locations"
+	"github.com/googleapis/api-linter/rules/internal/utils"
+	"github.com/jhump/protoreflect/desc"
+)
+
+// Custom methods should return a response message matching the RPC name,
+// with a Response suffix.
+var responseMessageName = &lint.MethodRule{
+	Name:   lint.NewRuleName(135, "response-message-name"),
+	OnlyIf: isCustomMethod,
+	LintMethod: func(m *desc.MethodDescriptor) []lint.Problem {
+		matchingResponseName := m.GetName() + "Response"
+		suggestion := matchingResponseName
+
+		got := m.GetOutputType().GetName()
+		want := stringset.New(matchingResponseName)
+
+		// Get the resource name from the first request message field, if it
+		// is a resource reference
```

Comment:
Interested to hear opinions about this approach - not sure if there are better ways of doing this.  Using `name` with a resource_reference seems to be convention for custom methods acting on a resource, but it's not spelled out in the AIPs.

Reply:
Yeah this will be a bit hand wavy.

What I think you are trying to do here (correct me if I'm wrong) is correlate a resource referenced in the request to the allowed resource message type name. There are a few issues with this (and I share them for education not trying to roast you 😅 ):
1. Exactly what you mentioned - there is no clearly defined pattern for finding *the* resource from the request that would be reasonable as a response message (could be field `name`, `parent`, `location`, `book`, etc.) - `name` is the _typical_ one for "should return the resource" but I'm not willing to bet this is always the case, ya know?
2. A resource being referenced in the request doesn't mean it is reasonable to return in the response i.e. we don't get enough signal from the field name alone to understand the semantics of the method
3. The reference may be a direct reference to a parent resource type rather than an indirect one via a child_type reference

It all kind of boils down to lately I have tried to avoid banking rules on heuristics informed by style i.e., field naming, since it's caused more friction than we'd like.

I'd separate the concerns and simply check if the response type is annotated as a resource (using `utils.IsResource`) as an adequate alternative to a `{MethodName}Response` response message name.

From there, we can add a separate rule to check "is this the right resource to return". This will give us a bit more flexibility. This is my rule of thumb no pun intended - rules should lint 1 thing and 1 thing only, anything extra is split into a separate rule.

WDYT? feel free to disagree/counter/ask questions, not trying to force this one

Reply:
> From there, we can add a separate rule to check "is this the right resource to return". This will give us a bit more flexibility. This is my rule of thumb no pun intended - rules should lint 1 thing and 1 thing only, anything extra is split into a separate rule.

I'm not entirely sure about this statement.  When you say "rules should lint 1 thing and 1 thing only, anything extra is split into a separate rule."  Can that be accomplished currently by this framework?

To me that would look something like this:

```
rules/aip0136/response_message_name_response_suffix.go
rules/aip0136/response_message_name_resource.go
```

But these are mutually exclusive such that one will always fail even when the other succeeds.  The logic for how we _lint the response message name_ has to happen in a single lint rule, right?

> I'd separate the concerns and simply check if the response type is annotated as a resource (using utils.IsResource) as an adequate alternative to a {MethodName}Response response message name.

I think this is a perfectly acceptable first pass, as it lints to the spirit of the rule.  Perhaps a second pass could to ascertain all of the models from the given request type and see if the model being returned is any of those.  Assuming it would be safe to guide developers to not response with implicitly modified resources derived from input.

Reply:
> But these are mutually exclusive such that one will always fail even when the other succeeds. The logic for how we lint the response message name has to happen in a single lint rule, right?

I'm kind of splitting hairs here, so 🐻 with me, but...what I was trying to suggest was:

* one rule enforcing that a custom method response message either: ends with `{Method}Response` OR is _any_ resource
* one rule enforcing that if a custom method response message is a resource, that it is *the right* resource

The second one is harder as we've discussed, and probably won't be _perfect_. WDYT?

> I think this is a perfectly acceptable first pass, as it lints to the spirit of the rule.

Agreed!

> Perhaps a second pass could to ascertain all of the models from the given request type and see if the model being returned is any of those. Assuming it would be safe to guide developers to not response with implicitly modified resources derived from input.

Yeah, exactly, and agreed, we'd run the risk of recommending the wrong resource 😅 --> developers dislike/lose trust in the linter

Reply:
Rethinking some of this @noahdietz, let me know if this sounds right to you.

According to https://google.aip.dev/136#resource-based-custom-methods

> The parameter for the resource's name must be called name, and be the only variable in the URI path.

What this tells me is that the only time a _resource_ could be returned is when `name` is a path param, which would mean it would exist on the request message type.

I think it would be safe to say:
- If the response is a resource type:
  - And if the `name` parameter is provided (this should be linted separately)
    - Then the name.resource_reference.type == response.resouce.type

Reply:
I'm nervous about releasing something that would allow this to pass, which could deviate from the spirit of the rules, and also as you mentioned, cause distrust in the tool

```proto
service Library {
  rpc ArchiveBook(ArchiveBookRequest) returns (Author);
}
```

Reply:
Good point! That was a (somewhat) recent change specifically to address the ambiguity of resource-based custom methods. It gives us a leg to stand on here.

Agreed your example would be scary to allow :)

> If the response is a resource type:
And if the name parameter is provided (this should be linted separately)
Then the name.resource_reference.type == response.resouce.type

Perfect, let's do it. Love the "this should be linted separately" 🙏🏻 

Thanks for working through that with me!

Reply:
<img src="https://media2.giphy.com/media/WrGiAHYhZZYZ2/giphy.gif"/>